app/assets/javascripts/oxalis/model/edge_collection.js

```diff
@@ -0,0 +1,179 @@
+    const outgoingEdges = this.outMap.has(nodeId) ? this.outMap.get(nodeId) : [];
+    const ingoingEdges = this.inMap.has(nodeId) ? this.inMap.get(nodeId) : [];
+
+    return outgoingEdges.concat(ingoingEdges);
+  }
+
+  addEdges(edges: Array<EdgeType>, mutate: boolean = false) {
+    // todo: find out whether the edge already exists?
+    const newOutgoingEdges = mutate ? this.outMap : this.outMap.clone();
+    const newIngoingEdges = mutate ? this.inMap : this.inMap.clone();
+    const newEdgeCount = this.edgeCount + edges.length;
+
+    for (const edge of edges) {
+      const outgoingEdges = this.outMap.has(edge.source) ? this.outMap.get(edge.source) : [];
+
+      const ingoingEdges = this.inMap.has(edge.target) ? this.inMap.get(edge.target) : [];
+
+      newOutgoingEdges.mutableSet(edge.source, outgoingEdges.concat(edge));
+      newIngoingEdges.mutableSet(edge.target, ingoingEdges.concat(edge));
+    }
+    if (mutate) {
+      this.edgeCount = newEdgeCount;
+      return this;
+    } else {
+      return EdgeCollection.loadFromMaps(newOutgoingEdges, newIngoingEdges, newEdgeCount);
+    }
+  }
+
+  addEdge(edge: EdgeType, mutate: boolean = false) {
+    return this.addEdges([edge], mutate);
+  }
+
+  removeEdge(edge: EdgeType) {
+    const outgoingEdges = this.outMap.has(edge.source) ? this.outMap.get(edge.source) : [];
+    const ingoingEdges = this.inMap.has(edge.target) ? this.inMap.get(edge.target) : [];
+
+    const newOutgoingEdges = this.outMap.set(
+      edge.source,
+      outgoingEdges.filter(e => e.target !== edge.target),
+    );
+    const newIngoingEdges = this.inMap.set(
+      edge.target,
+      ingoingEdges.filter(e => e.source !== edge.source),
+    );
+
+    return EdgeCollection.loadFromMaps(newOutgoingEdges, newIngoingEdges, this.edgeCount - 1);
+  }
+
+  map<T>(fn: (value: EdgeType) => T): Array<T> {
+    return this.asArray().map(fn);
+  }
+
+  *all(): Generator<EdgeType, void, void> {
+    for (const edgeArray of this.outMap.values()) {
+      for (const edge of edgeArray) {
+        yield edge;
+      }
+    }
+  }
+
+  asArray(): Array<EdgeType> {
+    return Array.from(this.all());
+  }
+
+  size(): number {
+    return this.edgeCount;
+  }
+
+  clone(): EdgeCollection {
+    const cloned = new EdgeCollection();
+    cloned.outMap = this.outMap.clone();
+    cloned.inMap = this.inMap.clone();
+    cloned.edgeCount = this.edgeCount;
+    return cloned;
+  }
+
+  static loadFromEdges(edges: Array<EdgeType>): EdgeCollection {
+    // Build up temporary data structures for fast bulk processing
+    const rawOutMap: { [number]: Array<EdgeType> } = {};
+    const rawInMap: { [number]: Array<EdgeType> } = {};
+    edges.forEach(edge => {
+      if (rawOutMap[edge.source]) {
+        rawOutMap[edge.source].push(edge);
+      } else {
+        rawOutMap[edge.source] = [edge];
+      }
+      // The doublyLinked flag determines (if true) that source AND target node should contain their connecting edge
```

Comment:
This comment no longer applies right? The EdgeCollection is always doublyLinked and if you call `getEdgesForNode` you get all edges of a node, whereas when you call `all` you get only the outgoing edges (I'm just reiterating this to make sure I understood this correctly :)).

Reply:
Yes, I deleted the comment now. Your reiteration is correct, but just to clarify: calling `all` will give you **all** edges for the given tree, since each outgoing edge is also an ingoing edge. Just to make sure, we are on the same page :)